docs/in-depth-guide-ingredient-groups.md

```diff
@@ -0,0 +1,125 @@
+# In Depth Guide: Ingredient Groups
+
+> **Draft**
+
+Sometimes a website will format the list ingredients into groups, where each group contains the ingredients needed for a particular aspect of the recipe. Recipe schema has no way to represent this grouping, so all of the ingredients are given in a single list and the groupings are lost.
+
+Some examples of recipes that have ingredient groups are :
+
+* https://cooking.nytimes.com/recipes/1024570-green-salad-with-warm-goat-cheese-salade-de-chevre-chaud
+* https://lifestyleofafoodie.com/air-fryer-frozen-french-fries
+* https://www.bbcgoodfood.com/recipes/monster-cupcakes
+
+Not all websites use ingredient groups, and those that do use ingredient groups will not use them for all recipes.
+
+This library allows a scraper to return the ingredients in the groups defined in the recipe, if the functionality is added to that scraper.
+
+## `ingredient_groups()`
+
+The `ingredient_groups()` function returns a list of `IngredientGroup` objects. Each `IngredientGroup` is a dataclass that represents a group of ingredients:
+
+```python
+@dataclass
+class IngredientGroup:
+    ingredients: List[str]
+    purpose: Optional[
+        str
+    ] = None  # this group of ingredients is {purpose} (e.g. "For the dressing")
+```
+
+The *purpose* is the ingredient group heading, such as *"For the dressing"*, *"For the sauce"* etc. The *ingredients* is the list of ingredients that comprise that group.
+
+This dataclass is defined in `_grouping_utils.py` and should be imported from there
+
+```pyth
+from ._grouping_utils import IngredientGroup
+```
+
+The `ingredient_groups()` function has a default implementation in the `AbstractScraper` class that returns a single `IngredientGroup` object with `purpose` set to `None` and `ingredients` set to the output from the scraper's `ingredients()` function.
+
+To add ingredient group support to a scraper, the `ingredient_group` function needs to be overridden in the scraper class. The are three main things to consider when implementing ingredient groups support:
+
+1. The groups and their contents are not present in the Recipe schema. They must be extracted from the recipe HTML.
+2. The ingredients found in `ingredients()` and `ingredient_groups()` must be the same. This may sound obvious but there can sometimes minor differences in the ingredients in the schema and the ingredients in the HTML.
+3. Not all recipes on a website will use ingredient groups, so the implementation must support recipes that do and recipes that don't have ingredient groups. For recipes that don't have ingredient groups, the output should be the same as default implementation (i.e. a single `IngredientGroup` with `purpose=None` and `ingredients=ingredients()`).
```

Comment:
Trying to make some of the wording here a bit more direct.  I don't think my suggestions are perfect, so I think some more edits may be needed:

```suggestion
Adding ingredient group support to a scraper involves overriding the `ingredient_group` function for it.  There are three important points to consider:

1. The schema.org Recipe format does not support groupings - so scraping HTML is going to be required in the implementation. 
2. The `group_ingredients` function requires that the schema.org ingredients list matches the ingredients found from the HTML - so it's important to make sure that these match.
3. Not all recipes on a website will use ingredient groups, so the implementation must degrade gracefully in cases where groupings aren't available. For recipes that don't have ingredient groups, the output should be the same as default implementation (i.e. a single `IngredientGroup` with `purpose=None` and `ingredients=ingredients()`).
```

Reply:
I don't agree with the change to the 2nd point. It isn't that the `group_ingredients` function needs them to match (in fact we added functionally to handle the cases where there isn't an exact match between the two sets of ingredients), it's more than the scraper should be presenting the same thing in two different ways: a single list or in groupings.

Reply:
That was original my thinking when writing this. 

However we added a test the `ScraperTest` base class that checks that both methods return the same ingredients, so the wording here should reflect that. I'll have a bit more of a think on how the phrase this.